### DISCUSSION TITLE
Unable to erase image slot using mcuboot serial command.

### DISCUSSION THREAD

**Discussion by CreativeIndustryAg** [zephyrproject-rtos/zephyr#50436] [Q&A] [answered]
Hello

Platform: NXP mimxrt1060
I'm trying to implement custom dfu, and it turned out i'm not able to write/erase anything to flash. 
I always get the same error (below), and what's more i'm event can't erase slot using mcuboot shell command:

uart:~$ mcuboot erase 1
[00:00:14.038,085] <dbg> flash_flexspi_nor: flash_flexspi_nor_write_enable: Enabling write
[00:00:14.046,813] <err> os: ***** USAGE FAULT *****
[00:00:14.052,398] <err> os:   Attempt to execute undefined instruction
[00:00:14.059,722] <err> os: r0/a1:  0x00000001  r1/a2:  0x80032174  r2/a3:  0x9e8f438e
[00:00:14.068,389] <err> os: r3/a4:  0x800024b8 r12/ip:  0x00000001 r14/lr:  0x00001251
[00:00:14.077,087] <err> os:  xpsr:  0x81000000
[00:00:14.082,275] <err> os: Faulting instruction address (r15/pc): 0x000005c6
[00:00:14.090,209] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:14.098,022] <err> os: Current thread: 0x800133c0 (shell_uart)
[00:00:14.105,072] <err> os: Halting system

When i use mcumgr via udp everything works.

Could you help me?

**Comment by DerekSnell** [comment]
Hi @CreativeIndustryAg ,

You mention using a custom DFU.  Are there some simple steps to replicate this issue on the mimxrt1060_evk board?

The usage fault is triggered by an undefined instruction.  That means the opcode fetched from memory is invalid.  And it reports the faulting instruction address is at 0x5C6, which is in ITCM RAM.  Since code must be copied to ITCM, is there valid executable code located there?  Have you already tried using a debugger and stepping through the code to see what leads up to this fault?

Best regards

**Comment by CreativeIndustryAg** [comment]
@DerekSnell 
Thanks for answer. :)
It turnout the problem disappear when i disable continuous reading  ADC (adc_context). 
DFU and ADC uses different tasks, and there is no data exchange between them. 

OZone debug call stack:
<!DOCTYPE html>

<html>
<head>

</head>

<body>


Level | Function | Stack Frame | Source | PC | Return Address | Stack Used
-- | -- | -- | -- | -- | -- | --
0 | BusFault Exception | 32 @ 0x80004840 |   | 0xCCCCCCCC | [0x80004858]: 0x00000CFC | 14080 | <BusFault Exception> | 32 @ 0x80004840 |   | 0xCCCCCCCC | [0x80004858]: 0x00000CFC | 1408
0 | flash_flexspi_nor_read_status | 24 @ 0x80004860 | flash_mcux_flexspi_nor.c:180:2 | 0x00000CF8 | [0x80004874]: 0x00000D36 | 1376
0 | flash_flexspi_nor_wait_bus_busy | 16 @ 0x80004878 | flash_mcux_flexspi_nor.c:299:3 | 0x00000D32 | [0x80004884]: 0x00000DC2 | 1352
0 | flash_flexspi_nor_write | 48 @ 0x80004888 | flash_mcux_flexspi_nor.c:372:3 | 0x00000DBE | [0x800048B4]: 0x6001BCEE | 1336
0 | flash_sync | 40 @ 0x800048B8 | /home/agr/zephyrproject/zephyr/include/zephyr/drivers/flash.h:205:2 | 0x6001BCEC | [0x800048DC]: 0x60041B84 | 1288
0 | stream_flash_buffered_write | 40 @ 0x800048E0 | stream_flash.c:199:3 | 0x60041B80 | [0x80004904]: 0x6003FA2E | 1248
0 | flash_img_buffered_write | 16 @ 0x80004908 | flash_img.c:44:2 | 0x6003FA2A | [0x80004914]: 0x60013964 | 1208
0 | dfu_upload_image | 32 @ 0x80004918 | dfu_img_utils.c:207:3 | 0x60013960 | [0x80004934]: 0x6003D638 | 1192
0 | parse_image_file | 8 @ 0x80004938 | dfu_https_ft.cpp:236:3 | 0x6003D634 | [0x8000493C]: 0x60013436 | 1160
0 | download_file | 1088 @ 0x80004940 | dfu_https_ft.cpp:215:5 | 0x60013434 | [0x80004D7C]: 0x60013608 | 1152
0 | download_new_image | 24 @ 0x80004D80 | dfu_https_ft.cpp:307:3 | 0x60013604 | [0x80004D94]: 0x60013712 | 64
0 | https_ft_dfu_task_entry_point | 32 @ 0x80004D98 | dfu_https_ft.cpp:347:9 | 0x6001370E | [0x80004DB4]: 0x6003DA14 | 40
0 | z_thread_entry | 8 @ 0x80004DB8 | thread_entry.c:36:2 | 0x6003DA12 | [0x80004DBC]: 0xAAAAAAAA | 8
0 | Top of stack - return address out of range: 0xAAAAAAAA |   |   |   |   |  


</body>

</html>

I still can't find where is the problem is and i can't reproduce it on samples.

**Comment by DerekSnell** [comment, marked as answer]
Hi @CreativeIndustryAg ,
I see the bus fault occurs in the FlexSPI driver when writing to flash.  If you are eXecuting-in-Place (XiP) from that flash while you are also writing to it, the FlexSPI driver should jump to code in ITCM, so it is not XIPing while erasing/writing the flash.  I wonder if your ADC reading or interrupt is trying to jump and XIP from flash during the flash write.  The app will not be able to execute from flash during those times, and that includes ISRs.  So that is where I would focus to troubleshoot this.

Let us know what you find
